scipy/stats/_mannwhitneyu.py

```diff
@@ -0,0 +1,430 @@
+    0.11134688653314041
+
+    If desired, we can disable the continuity correction to get a result
+    that agrees with that reported in [4]_.
+
+    >>> _, pnorm = mannwhitneyu(males, females, use_continuity=False,
+    ...                         method="asymptotic")
+    >>> print(pnorm)
+    0.0864107329737
+
+    Regardless of whether we perform an exact or asymptotic test, the
+    probability of the test statistic being as extreme or more extreme by
+    chance exceeds 5%, so we do not consider the results statistically
+    significant.
+
+    Suppose that, before seeing the data, we had hypothesized that females
+    would tend to be diagnosed at a younger age than males.
+    In that case, it would be natural to provide the female ages as the
+    first input, and we would have performed a one-sided test using
+    ``alternative = 'less'``: females are diagnosed at an age that is
+    stochastically less than that of males.
+
+    >>> res = mannwhitneyu(females, males, alternative="less", method="exact")
+    >>> print(res)
+    MannwhitneyuResult(statistic=3.0, pvalue=0.05555555555555555)
+
+    Again, the probability of getting a sufficiently low value of the
+    test statistic by chance under the null hypothesis is greater than 5%,
+    so we do not reject the null hypothesis in favor of our alternative.
+
+    If it is reasonable to assume that the means of samples from the
+    populations are normally distributed, we could have used a t-test to
+    perform the analysis.
+
+    >>> from scipy.stats import ttest_ind
+    >>> res = ttest_ind(females, males, alternative="less")
+    >>> print(res)
+    Ttest_indResult(statistic=-2.239334696520584, pvalue=0.030068441095757924)
+
+    Under this assumption, the *p*-value would be low enough to reject the
+    null hypothesis in favor of the alternative.
+
+    '''
+
+    x, y, use_continuity, alternative, axis_int, method = (
+        _mwu_input_validation(x, y, use_continuity, alternative, axis, method))
+
+    x, y, xy = _broadcast_concatenate(x, y, axis)
+
+    n1, n2 = x.shape[-1], y.shape[-1]
+
+    if method == "auto":
+        method = _mwu_choose_method(n1, n2, xy, method)
+
+    # Follows [2]
+    ranks = stats.rankdata(xy, axis=-1)  # method 2, step 1
+    R1 = ranks[..., :n1].sum(axis=-1)    # method 2, step 2
+    U1 = R1 - n1*(n1+1)/2                # method 2, step 3
+    U2 = n1 * n2 - U1                    # as U1 + U2 = n1 * n2
+
+    if alternative == "greater":
+        U, f = U1, 1  # U is the statistic to use for p-value, f is a factor
+    elif alternative == "less":
+        U, f = U2, 1  # Due to symmetry, use SF of U2 rather than CDF of U1
+    else:
+        U, f = np.maximum(U1, U2), 2  # multiply SF by two for two-sided test
+
+    if method == "exact":
+        p = _mwu_state.sf(U.astype(int), n1, n2)
+    elif method == "asymptotic":
+        z = _get_mwu_z(U, n1, n2, ranks, continuity=use_continuity)
+        p = stats.norm.sf(z)
+    p *= f
```

Comment:
I thin the code is easier to understand if you write: 

```
If alternative == ...:
    p = 2 * p

Reply:
But if we were to repeat the branching structure `if alternative == ...` here, another reviewer might make an equally valid comment about D.R.Y. : )

Here's one thing I could do: in the `if method == ...` above, I could define the appropriate survival function to use based on the provided `method`, e.g.:

```
    if method == "exact":
        def sf(U, n1, n2):
            return _mwu_state.sf(U.astype(int), n1, n2)
    elif method == "asymptotic":
        def sf(U, n1, n2):
            z = _get_mwu_z(U, n1, n2, ranks, continuity=use_continuity)
            return stats.norm.sf(z)
```
(Or I could branch inside the function definition, if you prefer.)

Remove the `if alternative == ...` above and have:

```
```suggestion
    if alternative == "greater":
        U = U1  # U is the statistic to use for p-value
    elif alternative == "less":
        U = U2  # Due to symmetry, use SF of U2 rather than CDF of U1
    else:
        U = np.maximum(U1, U2)  # multiply SF by two for two-sided test
    p = sf(U, n1, n2)
```

Let me know if you'd prefer that.

Reply:
it is a minor point, it just confused me when I saw the multiplication of p by f and the explanation of f seems a bit vague at the moment: `f is a factor`. what about:

```
# set statistic to use for p-value depending on alternative
# note that p-value needs to be multiplied by f=2 in the two-sided case
f = 1
if alternative == "greater":
    U = U1
elif alternative == "less":
    U = U2  # Due to symmetry, use SF of U2 rather than CDF of U1
else:
    U, f = np.maximum(U1, U2), 2 
```

I hope it is easier to understand but feel free to keep the current version

Reply:
i updated the comment to keep my proposal closer to the original version

Reply:
Suggested [here](https://github.com/scipy/scipy/pull/4933/files#r620839786); please commit if it is preferred.